### ISSUE TITLE
CROSSTOOL: "with_feature" bug in "tool" of "action_config"

### ISSUE THREAD

**Issue by rongjiecomputer** [bazelbuild/bazel#5504]
### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Copy a full `CROSSTOOL` and change the following `action_config`.

```
  action_config {
    config_name: 'c++-link-executable'
    action_name: 'c++-link-executable'
    # Change the tool_paths to something presents in your system, the point is to see which tool
    # get selected.
    # Tool 1
    tool {
      tool_path: 'C:/Users/jie/Tools/LLVM/bin/lld-link.exe'
      with_feature {
        feature: 'full_llvm_toolchain'
      }
    }
    # Tool 2
    tool {
      tool_path: 'C:/Users/jie/Tools/VS2017/VC/Tools/MSVC/14.14.26428/bin/HostX64/x64/link.exe'
      # section 1
      #with_feature {
      #  not_feature: 'full_llvm_toolchain'
      #}
    }
    ...
```

Try to use this `CROSSTOOL` to build one binary (run with `-s` to see which tool get selected). Run `bazel clean` after each run.

1. `bazel build :binary -s`
   - Expected: second tool get selected.
   - Result: second tool get selected. (OK)
1. `bazel build --features=full_llvm_toolchain :binary -s`
   - Expected: first tool get selected.
   - Result: second tool get selected.
1. `bazel build --features=full_llvm_toolchain :binary -s` (uncomment section 1)
   - Expected: first tool get selected.
   - Result: second tool get selected.
1. `bazel build --features=full_llvm_toolchain :binary -s` (change `feature` in tool 1 to `not_feature`)
   - Expected: throw error (since both tools disallow `full_llvm_toolchain`).
   - Result: first tool get selected.

Note that `with_feature` with complex `feature` and `not_feature` logic works fine for `flag_set`.

### What operating system are you running Bazel on?

Windows 10 x64

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

Bazel built fresh from master branch

###  Have you found anything relevant by searching the web?

From https://github.com/bazelbuild/bazel/blob/master/src/main/protobuf/crosstool_config.proto#L293L297

**Comment by lfpino** [comment]
Assigning to @mhlopko for triage.

**Comment by mhlopko** [comment]
Nice catch. The bug will be somewhere here: https://source.bazel.build/bazel/+/6b8c7ca48e9e48a9906c031346dc4a924b7ef559:src/main/java/com/google/devtools/build/lib/rules/cpp/CcToolchainFeatures.java;l=988

Want to give it a try?

**Comment by rongjiecomputer** [comment]
@mhlopko I spent a few hours following the logic with source.bazel.build's cross referencing, the logic is all correct. It was only when I read the following test that made me realize what I did wrong.

https://source.bazel.build/bazel/+/b9166943e5dc23f7ce8ef2dce3cb98e5ce26e50b:src/test/java/com/google/devtools/build/lib/rules/cpp/CcToolchainFeaturesTest.java;l=1440

Root cause: I forgot to declare an empty `feature { name: "full_llvm_toolchain" }`. Adding this makes the `CROSSTOOL` works exactly as expected.

Apparently my custom `CROSSTOOL` file has outgrown my mind capacity. 😆
